<div style="
    padding: 20px; 
    text-align: center; 
    border-radius: 10px; 
    color: #87CEEB; 
    font-size: 24px; 
    text-shadow: 0 0 10px #87CEEB, 0 0 20px #87CEEB, 0 0 30px #87CEEB, 0 0 40px #87CEEB;
">
    <h1> ☀️ Top 100 Cities Weather Analysis 🌧️</h1>
    <h5> 🌍 Uncovering Global Weather Patterns with Data 📊</h5>
</div>


<div >
  <img src="https://res.cloudinary.com/dhditogyd/image/upload/v1726342283/weather_pqpibg.webp" width="100%" height=" 80%;"/>
</div>

---



Author: [Muhammad Faizan](https://www.linkedin.com/in/mrfaizanyousaf/)

<div >
  <img src="https://res.cloudinary.com/dk0hm8t43/image/upload/v1723543716/Faizan/cxhm4yhcpmponfjgjfct.jpg" width="20%" height=" 20%;"/>
</div>

<br>
<br>

<div style="
    padding: 20px; 
    color: #87CEEB; 
    font-size: 20px; 
    text-shadow: 0 0 10px #87CEEB, 0 0 20px #87CEEB, 0 0 30px #87CEEB, 0 0 40px #87CEEB;
">
    <h3> Muhammad Faizan </h3>
</div>

🎓 **3rd Year BS Computer Science** student at the **University of Agriculture, Faisalabad**  
💻 Enthusiast in **Machine Learning, Data Engineering, and Data Analytics**


### 🌐 Connect with Me

[Kaggle](https://www.kaggle.com/faizanyousafonly/) | [LinkedIn](https://www.linkedin.com/in/mrfaizanyousaf/) | [GitHub](https://github.com/faizan-yousaf/)  




### 💬 Contact Me
- **Email:** faizanyousaf815@gmail.com
- **WhatsApp:** [+92 306 537 5389](https://wa.me/923065375389)


🔗 **Let’s Collaborate:**  
I'm always open to queries, collaborations, and discussions. Let's build something amazing together!


<div style="
    padding: 20px;
    text-align: center;
    border-radius: 10px;
    color: #00FFFF;
    font-size: 24px;
    text-shadow: 0 0 10px #87CEEB, 0 0 20px #87CEEB, 0 0 30px #87CEEB, 0 0 40px #87CEEB;
">
    <h1> Meta-Data (About Dataset) ⛅</h1>
</div>

## Importance of Weather Data
Weather data plays a crucial role in various fields such as `agriculture`, `transportation`, `energy`, and `tourism`. Farmers rely on weather forecasts to plan planting and harvesting schedules. Airlines and shipping companies use weather data to optimize routes and ensure passenger safety. Energy companies use weather forecasts to predict demand and manage resources efficiently. Additionally, weather data is essential for tourism, as travelers often base their destination choices and activities on current and forecasted weather conditions.

> Overall, **accurate** and **reliable** weather data is vital for making informed decisions and mitigating risks in various industries and everyday life.

---

## Background
My friend, a travel agent, often faces challenges when presenting travel destinations to clients. One common question from clients is about the **weather conditions** of their desired destinations. To address this, my friend asked me to collect data on the **top 100 tourist destination cities worldwide** for `April 28th, 2024`.

This data will help him provide better presentations to his clients by offering **detailed insights** into the weather conditions of these cities on that specific date. With this information readily available, my friend can enhance his client consultations and ensure that travelers have a more enjoyable and hassle-free experience during their trips.

---

## Introduction of Dataset
This dataset provides **temperature data** for the top 100 cities worldwide as of `April 28th, 2024`. It consists of seven columns, each offering specific information to understand the weather conditions in different cities. The dataset includes the following columns:

### 1. `City`
The name of the city in English language. This column provides the official or commonly used name of each city. City names are essential for identifying the geographical location associated with the weather data.

### 2. `Temperature (°C)`
The temperature recorded for each city on `April 28th, 2024`. Temperature is measured in degrees Celsius (`°C`) and represents the average atmospheric temperature at the specified location on the given date. It is an essential parameter in weather analysis, indicating the warmth or coldness of the air.

### 3. `Wind Speed (m/s)`
The average wind speed measured at each location in `meters per second (m/s)`. Wind speed is a measure of the movement of air molecules and is an important factor in weather prediction and analysis.

### 4. `Latitude (°)`
The latitude coordinates of each city, expressed in degrees. Latitude measures the north-south position of a location on the Earth's surface relative to the equator. Positive values represent locations in the northern hemisphere, while negative values represent locations in the southern hemisphere. Latitude ranges from `-90°` (South Pole) to `+90°` (North Pole), with `0°` at the equator.

### 5. `Longitude (°)`
The longitude coordinates of each city, expressed in degrees. Longitude measures the east-west position of a location on the Earth's surface relative to the Prime Meridian. Positive values represent locations east of the Prime Meridian, while negative values represent locations west of it. Longitude ranges from `-180°` (West) to `+180°` (East), with `0°` at the Prime Meridian.

### 6. `Description`
This column provides details about the prevailing weather conditions in each city on `April 28th, 2024`. It includes descriptions such as `clear sky`, `scattered clouds`, `broken clouds`, `overcast clouds`, and `few clouds`, indicating the level of cloud cover and sky visibility at the time of observation.

### 7. `Country`
This column contains the name of the country to which each city belongs. It provides information about the city with its respective country.

### Acknowledgements
#### Creator:

* Danish Ammar, Data Scientist, Kaggle Dataset Grand Master, Islamabad Capital Territory, Pakistan. [Link](https://www.kaggle.com/danishammar)
* Dataset link: [Top 100 Cities Weather](https://www.kaggle.com/datasets/danishammar/top-100-cities-weather-dataset/data)


<div style="
    padding: 20px;
    text-align: center;
    border-radius: 10px;
    color: #00FFFF;
    font-size: 24px;
    text-shadow: 0 0 10px #87CEEB, 0 0 20px #87CEEB, 0 0 30px #87CEEB, 0 0 40px #87CEEB;
">
   <h1> Workflow of this project: 🌍📊</h1>
</div>

### 1. `Data Scraping (Data Collection using OpenWeather API)`
   - Collect real-time weather data for cities in Pakistan using OpenWeather API.
   - Gather details such as temperature, humidity, wind speed, cloudiness, and description.
   - Store the data in a structured format (e.g., CSV or DataFrame).

### 2. `Data Exploration`
   - Load the dataset and inspect the structure.
   - Perform initial exploration to understand features.
   - Analyze summary statistics for both numerical and categorical variables.

### 3. `Data Cleaning`
   - Handle missing data (e.g., fill missing values, drop duplicates).
   - Convert columns (e.g., `Sunrise`, `Sunset`) to appropriate data types.
   - Normalize column names for consistency.
   - Remove unnecessary columns or rows containing errors or null values.

### 4. `Exploratory Data Analysis (EDA)`
   - Analyze the distribution of numerical variables (e.g., temperature, humidity).
   - Explore correlations between different features.
   - Investigate outliers and detect patterns in the dataset.
   - Group cities and compare weather conditions.

### 5. `Data Visualization Using Seaborn and Matplotlib`
   - Create visualizations such as histograms, scatterplots, boxplots, and heatmaps.
   - Visualize weather condition trends across cities.
   - Show relationships between variables (e.g., temperature vs. humidity, cloudiness vs. wind speed).


### Import Libraries: 🛠

Let's start the project by importing all the libraries that we will use in this project.

In [20]:
# import libraries:

# 1. for sending HTTP requests:
import requests

# 2. for manipulating dates and times:
from datetime import datetime

# 3. to handel the data:
import numpy as np
import pandas as pd

# 4. to visualize the data:
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
%matplotlib inline

# 5. to ignore the warnings:
import warnings
warnings.filterwarnings("ignore")

print("Libraries have been loaded successfully")

# 6. Display all rows and columns:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)


Libraries have been loaded successfully


#### But why these libraries? 🤔


* `requests:` For making HTTP requests to the OpenWeather API.

* `datetime:` For working with dates and times.

* `pandas:` For data manipulation and analysis.

* `numpy:` For numerical computations.
 
* `matplotlib & seaborn:` For data visualization.
 
* `%matplotlib inline:` To display plots directly in the notebook.

#### Setting up the Visualization Theme: 🐱‍🏍✨
(I love to work on my datasets with a dark theme, it's more fun and cool. So, I will set the dark theme for my visualizations.) 🎇

In [21]:
# Set the dark gray background theme for visualizations
plt.style.use('dark_background')

# Adjust background color to dark gray
plt.rcParams['figure.facecolor'] = '#2e2e2e'
plt.rcParams['axes.facecolor'] = '#2e2e2e'
plt.rcParams['axes.edgecolor'] = '#2e2e2e'
plt.rcParams['savefig.facecolor'] = '#2e2e2e'

# color palette with cloudy and rainy effect
weather_palette = sns.color_palette(["#a9a9a9", "#c0c0c0", "#b0e0e6", "#87ceeb", "#4682b4"])

# Apply the custom palette to Seaborn
sns.set_palette(weather_palette)

## Let's start our project step by step:
### **Part 1: Data Collection:**

#### 1. `Data Scraping (Data Collection using OpenWeather API)`
   - Collect real-time weather data for cities in Pakistan using OpenWeather API.
   - Gather details such as temperature, humidity, wind speed, cloudiness, and description.
   - Store the data in a structured format (e.g., CSV or DataFrame).


#### 2. Use the OpenWeather API to collect real-time weather data for TOP 100 cities of the World.

In [22]:
# first, we will load the list of cities from a CSV file

cities = pd.read_csv('top100cities_weather_data.csv')  
cities.tail()

City  Temperature (Celsius)  Wind Speed (m/s)  Latitude  Longitude  \
95  Punta Cana                  26.87              6.69   18.5818   -68.4043   
96      Guilin                  24.06              0.97   25.2819   110.2864   
97       Hanoi                  28.00              2.49   21.0245   105.8412   
98       Cairo                  19.42              4.63   30.0626    31.2497   
99      Muscat                  28.01              2.76   23.6139    58.5922   

      Description             Country  
95     few clouds  Dominican Republic  
96      clear sky               China  
97      clear sky             Vietnam  
98      clear sky               Egypt  
99  broken clouds                Oman

In [23]:
# Your API key
api_key = '65190f9772626f8c7f3e0b041a41dd07'

# Define an empty list to store the results
weather_data = []

# Loop through each city and its corresponding country in the DataFrame
for index, row in cities.iterrows():
    city = row['City']
    country = row['Country']  
    # Construct the API request URL
    url = f'http://api.openweathermap.org/data/2.5/weather?q={city}&appid={api_key}&units=metric'

    # Make the request to the API
    response = requests.get(url)

    # Check if the request was successful
    if response.status_code == 200:
        data = response.json()  # Convert response to JSON format

        # Extract weather data
        temp = data['main']['temp']
        feels_like = data['main']['feels_like']
        temp_min = data['main']['temp_min']
        temp_max = data['main']['temp_max']
        humidity = data['main']['humidity']
        pressure = data['main']['pressure']
        wind_speed = data['wind']['speed']
        desc = data['weather'][0]['description']
        #country = data['sys']['country']
        clouds = data['clouds']['all']
        lon = data['coord']['lon']
        lat = data['coord']['lat']
        sunrise_unix = data['sys']['sunrise']
        sunset_unix = data['sys']['sunset']

        # Convert Unix timestamps to human-readable format
        sunrise = datetime.fromtimestamp(sunrise_unix).strftime('%I:%M %p')
        sunset = datetime.fromtimestamp(sunset_unix).strftime('%I:%M %p')

        # Append the data to the list
        weather_data.append({
            'Country': country,
            'City': city,
            'Temperature': temp,
            'Feels_Like': feels_like,
            'Min_Temp': temp_min,
            'Max_Temp': temp_max,
            'Humidity': humidity,
            'Pressure': pressure,
            'Wind_Speed': wind_speed,
            'Cloudiness': clouds,
            'Description': desc.capitalize(),
            'Longitude': lon,
            'Latitude': lat,
            'Sunrise': sunrise,
            'Sunset': sunset
        })

    else:
        # If the request fails, append an error message for that city
        weather_data.append({
            'City': city,
            'Error': f'Error fetching data: {response.status_code}'
        })

# Convert the list of weather data into a pandas DataFrame
weather_df = pd.DataFrame(weather_data)

In [24]:
# save the dataset that is fetched from the API to a CSV file:

weather_df.to_csv('top_100_cities_of_the_world_weather_data.csv', index=False)

### **Part 2: Data Collection:**

#### 1. Load the Dataset 

In [25]:
df = pd.read_csv('top_100_cities_of_the_world_weather_data.csv')

#### 2. Let's have a Sneek peak at the data: 👀

First 10 rows:

In [26]:
print("The first 10 rows of the dataset are shown below:\n") 
df.head(10)


The first 10 rows of the dataset are shown below:



Country           City  Temperature  Feels_Like  Min_Temp  \
0                France          Paris        19.58       18.79     18.45   
1  United Arab Emirates          Dubai        36.54       41.84     35.14   
2                 Spain         Madrid        28.82       27.24     28.34   
3                 Japan          Tokyo        26.64       26.64     25.34   
4           Netherlands      Amsterdam        18.60       18.27     17.69   
5               Germany         Berlin        18.05       17.29     16.67   
6                 Italy           Rome        23.34       23.66     21.93   
7         United States  New York City        22.35       22.55     20.75   
8                 Spain      Barcelona        26.53       26.53     25.14   
9        United Kingdom         London        19.34       18.92     18.12   

   Max_Temp  Humidity  Pressure  Wind_Speed  Cloudiness  \
0     20.70        46      1026        2.57           0   
1     36.96        44      1002        8.23           0   
2     30.09        15      1017        2.06           0   
3     28.12        89      1013        4.12          75   
4     19.38        67      1025        4.63          40   
5     18.88        53      1020        9.83          20   
6     24.18        74      1018        4.63         100   
7     24.06        73      1028        6.17           0   
8     29.52        37      1016        3.13           0   
9     20.27        61      1026        4.63         100   

                   Description  Longitude  Latitude   Sunrise    Sunset  
0                    Clear sky     2.3488   48.8534  10:27 AM  11:04 PM  
1                    Clear sky    55.3047   25.2582  07:04 AM  07:23 PM  
2                    Clear sky    -3.7026   40.4165  10:55 AM  11:24 PM  
3  Light intensity shower rain   139.6917   35.6895  01:23 AM  01:49 PM  
4             Scattered clouds     4.8897   52.3740  10:15 AM  10:56 PM  
5                   Few clouds    13.4105   52.5244  09:41 AM  10:22 PM  
6              Overcast clouds   -85.1647   34.2570  04:24 PM  04:47 AM  
7                    Clear sky   -74.0060   40.7143  03:37 PM  04:05 AM  
8                    Clear sky     2.1590   41.3888  10:32 AM  11:01 PM  
9              Overcast clouds    -0.1257   51.5085  10:35 AM  11:15 PM

Last 10 rows:

In [27]:
print("\n\nThe last 10 rows of the dataset are shown below:\n")
df.tail(10)



The last 10 rows of the dataset are shown below:



Country        City  Temperature  Feels_Like  Min_Temp  \
90             Morocco   Marrakech        36.04       36.22     36.04   
91           Lithuania     Vilnius        23.49       23.36     23.18   
92              Turkey       Mugla        23.92       23.31     23.92   
93               China      Zhuhai        29.28       36.28     28.84   
94               India      Mumbai        26.99       29.53     26.99   
95  Dominican Republic  Punta Cana        28.87       35.87     28.87   
96               China      Guilin        24.06       24.79     24.06   
97             Vietnam       Hanoi        29.00       36.00     29.00   
98               Egypt       Cairo        34.42       36.16     34.12   
99                Oman      Muscat        36.01       42.72     36.01   

    Max_Temp  Humidity  Pressure  Wind_Speed  Cloudiness      Description  \
90     39.08        30      1010        5.14           0        Clear sky   
91     23.49        56      1017        3.60           0        Clear sky   
92     23.92        36      1007        4.69           0        Clear sky   
93     30.40        84      1003        2.06          20       Few clouds   
94     26.99        78      1008        1.54          75             Mist   
95     28.87        89      1014        4.12          75    Broken clouds   
96     24.06        87      1006        1.59          68    Broken clouds   
97     29.00        89      1003        2.56         100  Overcast clouds   
98     34.42        40      1008        6.69          75    Broken clouds   
99     36.01        49      1002        2.87          90  Overcast clouds   

    Longitude  Latitude   Sunrise    Sunset  
90    -8.0083   31.6315  11:16 AM  11:38 PM  
91    25.2798   54.6892  08:52 AM  09:36 PM  
92    28.3557   37.2303  08:48 AM  09:14 PM  
93   113.5678   22.2769  03:12 AM  03:29 PM  
94    72.8479   19.0144  05:56 AM  06:11 PM  
95   -68.4043   18.5818  03:21 PM  03:36 AM  
96   110.2864   25.2819  03:24 AM  03:43 PM  
97   105.8412   21.0245  03:43 AM  04:00 PM  
98    31.2497   30.0626  08:39 AM  09:00 PM  
99    58.5922   23.6139  06:52 AM  07:09 PM

#### 3. Check the Structure of the Dataset


In [28]:
# number of rows and columns in the dataset

print(f"\n\nThe dataset contains {df.shape[0]} rows and {df.shape[1]} columns.")



The dataset contains 100 rows and 15 columns.


#### 4. Check data types of all the columns:

In [29]:
# data types of each column in the dataset

print("\n\nThe data types of each column in the dataset are as follows:\n")
df.dtypes



The data types of each column in the dataset are as follows:



Country         object
City            object
Temperature    float64
Feels_Like     float64
Min_Temp       float64
Max_Temp       float64
Humidity         int64
Pressure         int64
Wind_Speed     float64
Cloudiness       int64
Description     object
Longitude      float64
Latitude       float64
Sunrise         object
Sunset          object
dtype: object

#### 5. Check the Statistical Summary:

In [30]:
# summary statistics of the dataset

print("\n\nThe summary statistics of the dataset are as follows:\n")
df.describe().T.round(2)





The summary statistics of the dataset are as follows:



count     mean    std      min      25%      50%      75%  \
Temperature  100.0    24.35   7.03     8.58    18.68    25.59    28.83   
Feels_Like   100.0    25.59   8.91     3.08    18.34    25.94    32.57   
Min_Temp     100.0    23.49   7.35     6.99    17.66    24.94    28.12   
Max_Temp     100.0    25.23   6.88    10.26    20.22    26.96    29.42   
Humidity     100.0    63.18  21.78    10.00    46.75    68.00    82.00   
Pressure     100.0  1012.67   7.24  1001.00  1007.75  1012.00  1017.00   
Wind_Speed   100.0     3.81   2.66     0.45     2.06     3.09     4.80   
Cloudiness   100.0    37.22  36.57     0.00     0.00    20.00    75.00   
Longitude    100.0    18.55  75.88  -157.86    -8.74    18.55    74.15   
Latitude     100.0    31.25  19.45   -34.61    23.49    35.17    43.83   

                 max  
Temperature    40.08  
Feels_Like     42.72  
Min_Temp       40.08  
Max_Temp       40.08  
Humidity       97.00  
Pressure     1028.00  
Wind_Speed     16.99  
Cloudiness    100.00  
Longitude     151.21  
Latitude       59.44

#### Univariate Analysis: 

In [31]:
# let's make a for loop to get the minimum and maximum value of each numerical column in the dataset:

for num_col in df.select_dtypes(include=np.number).columns:
    min_val = df[num_col].min()
    max_val = df[num_col].max()
    print(f"The minimum value of {num_col} is {min_val} \nThe maximum value of {num_col} is {max_val}.")
    print("\n")

The minimum value of Temperature is 8.58 
The maximum value of Temperature is 40.08.


The minimum value of Feels_Like is 3.08 
The maximum value of Feels_Like is 42.72.


The minimum value of Min_Temp is 6.99 
The maximum value of Min_Temp is 40.08.


The minimum value of Max_Temp is 10.26 
The maximum value of Max_Temp is 40.08.


The minimum value of Humidity is 10 
The maximum value of Humidity is 97.


The minimum value of Pressure is 1001 
The maximum value of Pressure is 1028.


The minimum value of Wind_Speed is 0.45 
The maximum value of Wind_Speed is 16.99.


The minimum value of Cloudiness is 0 
The maximum value of Cloudiness is 100.


The minimum value of Longitude is -157.8583 
The maximum value of Longitude is 151.2073.


The minimum value of Latitude is -34.6132 
The maximum value of Latitude is 59.437.




#### 6. Check for Unique Values in Categorical Columns


In [32]:
# I'll use a for loop and will print the unique values of all categorical columns:

for col in df.select_dtypes(include='object').columns:
    print(f"\nNumber of unique values in the column '{col}' are: {df[col].nunique()}")
    print(f"\n\nUnique values in the column '{col}' are:\n")
    print(df[col].unique())
    
    print("---------------------------------------------------------------------------------------------")


Number of unique values in the column 'Country' are: 47


Unique values in the column 'Country' are:

['France' 'United Arab Emirates' 'Spain' 'Japan' 'Netherlands' 'Germany'
 'Italy' 'United States' 'United Kingdom' 'Singapore' 'South Korea'
 'Ireland' 'China' 'Austria' 'Portugal' 'Czech Republic' 'Australia'
 'Turkey' 'Taiwan' 'Canada' 'Greece' 'Switzerland' 'Thailand' 'Malaysia'
 'Sweden' 'Belgium' 'Israel' 'Poland' 'Denmark' 'Hungary' 'Brazil'
 'Mexico' 'Pakistan' 'Qatar' 'Argentina' 'Peru' 'India' 'Estonia'
 'Saudi Arabia' 'Vietnam' 'Chile' 'Georgia' 'Morocco' 'Lithuania'
 'Dominican Republic' 'Egypt' 'Oman']
---------------------------------------------------------------------------------------------

Number of unique values in the column 'City' are: 100


Unique values in the column 'City' are:

['Paris' 'Dubai' 'Madrid' 'Tokyo' 'Amsterdam' 'Berlin' 'Rome'
 'New York City' 'Barcelona' 'London' 'Singapore' 'Munich' 'Milan' 'Seoul'
 'Dublin' 'Osaka' 'Hong Kong' 'Vienna' 'Los Ange

In [33]:
# Unique weather descriptions
df['Description'].value_counts()

Description
Clear sky                      34
Few clouds                     16
Broken clouds                  15
Overcast clouds                14
Scattered clouds               12
Light rain                      3
Mist                            2
Light intensity shower rain     1
Light intensity drizzle         1
Smoke                           1
Moderate rain                   1
Name: count, dtype: int64

In [34]:
# Unique countries:
df['Country'].value_counts()

Country
United States           9
China                   8
Spain                   5
Italy                   5
Japan                   5
Greece                  4
Germany                 3
Mexico                  3
Canada                  3
Turkey                  3
United Arab Emirates    3
Portugal                2
France                  2
South Korea             2
United Kingdom          2
Poland                  2
Saudi Arabia            2
India                   2
Malaysia                2
Thailand                2
Brazil                  2
Israel                  2
Australia               2
Vietnam                 2
Ireland                 1
Singapore               1
Netherlands             1
Austria                 1
Switzerland             1
Czech Republic          1
Taiwan                  1
Hungary                 1
Denmark                 1
Sweden                  1
Belgium                 1
Argentina               1
Qatar                   1
Peru                    1
Paki

> #### Seems like we have to convert the data types of some columns:

But wait, first let's have a look on all the data types:


1. **`int`**: Represents integer numbers without decimal points.
   ```python
   num = 42
   ```

2. **`float`**: Represents floating-point numbers with decimal points.
   ```python
   num = 3.14
   ```

3. **`str`**: Represents strings of text.
   ```python
   text = "Hello, DATA SEEKHO TEAM :) !"
   ```

4. **`bool`**: Represents Boolean values (`True` or `False`).
   ```python
   flag = True
   ```

5. **`list`**: Represents an ordered, mutable collection of items.
   ```python
   items = [1, 2, 3, 4]
   ```

6. **`tuple`**: Represents an ordered, immutable collection of items.
   ```python
   coords = (10, 20)
   ```

7. **`set`**: Represents an unordered collection of unique items.
   ```python
   unique_items = {1, 2, 3, 4}
   ```

8. **`dict`**: Represents a collection of key-value pairs.
   ```python
   person = {'name': 'Muhammad Faizan', 'age': 21}
   ```

9. **`NoneType`**: Represents the absence of a value or a null value.
   ```python
   value = None
   ```

10. **`bytes`**: Represents immutable sequences of bytes.
    ```python
    data = b'hello'
    ```

11. **`bytearray`**: Represents mutable sequences of bytes.
    ```python
    data = bytearray([65, 66, 67])
    ```


> BUT, there is a difference between `str` and `category` data type, let's see the difference first:

In Python/Pandas, `str` and `category` are two different data types used to handle textual data, each with its own characteristics and use cases. Here’s a detailed comparison:

---
* `str` Data Type
---

1. **Definition**:
   - Represents strings of text.
   - Standard Python `str` type, used to store textual data.

2. **Storage**:
   - Each unique string is stored as is.
   - Memory usage can be higher, especially with a large number of unique strings.

3. **Operations**:
   - Supports all standard string operations and methods.
   - Slower for operations involving large datasets with many unique strings, due to the lack of optimization for categorical data.

4. **Use Case**:
   - Ideal for columns with a small number of unique strings or when the exact text value is important.

5. **Example**:
   ```python
   df['text_column'] = df['text_column'].astype(str)
   ```
---
* `category` Data Type
---

1. **Definition**:
   - A Pandas-specific data type optimized for categorical data.
   - Represents a fixed number of possible values (categories).

2. **Storage**:
   - Internally stores data as integer codes and a separate array of category labels.
   - More memory-efficient, especially with many repeated values.

3. **Operations**:
   - Limited string operations compared to `str`.
   - Faster for operations involving large datasets with repeating categories due to the use of integer codes.
   - Supports category-specific operations like reordering, adding, or removing categories.

4. **Use Case**:
   - Ideal for columns with a limited set of distinct values that repeat often (e.g., gender, color, or category labels).
   - Useful for improving performance and memory usage in large datasets with repetitive text.

5. **Example**:
   ```python
   df['category_column'] = df['category_column'].astype('category')
   ```

**Summary**

- Use `str` when you need full string functionality and when dealing with text data where every unique string matters.
- Use `category` when you have a column with a limited number of unique values that repeat often, as it provides efficiency improvements in both memory and performance. 

**Choosing the right type depends on the specific needs of your data and the operations you need to perform.**

### Data Consistency:

In [35]:
# make sure that the each column has correct data type:

df['Country'] = df['Country'].astype('category')
df['City'] = df['City'].astype('str')
df['Description'] = df['Description'].astype('category')
df['Temperature'] = df['Temperature'].astype('float')
df['Feels_Like'] = df['Feels_Like'].astype('float')
df['Min_Temp'] = df['Min_Temp'].astype('float')
df['Max_Temp'] = df['Max_Temp'].astype('float')
df['Humidity'] = df['Humidity'].astype('int')
df['Pressure'] = df['Pressure'].astype('int')
df['Wind_Speed'] = df['Wind_Speed'].astype('float')
df['Cloudiness'] = df['Cloudiness'].astype('int')
df['Longitude'] = df['Longitude'].astype('float')
df['Latitude'] = df['Latitude'].astype('float')
df['Sunrise'] = df['Sunrise'].astype('str')
df['Sunset'] = df['Sunset'].astype('str')


In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   Country      100 non-null    category
 1   City         100 non-null    object  
 2   Temperature  100 non-null    float64 
 3   Feels_Like   100 non-null    float64 
 4   Min_Temp     100 non-null    float64 
 5   Max_Temp     100 non-null    float64 
 6   Humidity     100 non-null    int64   
 7   Pressure     100 non-null    int64   
 8   Wind_Speed   100 non-null    float64 
 9   Cloudiness   100 non-null    int64   
 10  Description  100 non-null    category
 11  Longitude    100 non-null    float64 
 12  Latitude     100 non-null    float64 
 13  Sunrise      100 non-null    object  
 14  Sunset       100 non-null    object  
dtypes: category(2), float64(7), int64(3), object(3)
memory usage: 12.3+ KB


In [37]:
df['City'].astype(str).str.len().max()

np.int64(17)

In [38]:
df['City'].value_counts() 
# it was obvious that we had 100 uniques cities :)

City
Paris                1
Dubai                1
Madrid               1
Tokyo                1
Amsterdam            1
Berlin               1
Rome                 1
New York City        1
Barcelona            1
London               1
Singapore            1
Munich               1
Milan                1
Seoul                1
Dublin               1
Osaka                1
Hong Kong            1
Vienna               1
Los Angeles          1
Lisbon               1
Prague               1
Sydney               1
Istanbul             1
Melbourne            1
Orlando              1
Frankfurt            1
Kyoto                1
Taipei               1
Florence             1
Toronto              1
Athens               1
Zurich               1
Bangkok              1
Las Vegas            1
Miami                1
Kuala Lumpur         1
Venice               1
Abu Dhabi            1
Stockholm            1
Brussels             1
Tel Aviv             1
San Francisco        1
Shanghai             1
Warsaw

Missing values:

In [39]:
# number of missing values in each column of the dataset

print("\n\nThe number of missing values in each column of the dataset are as follows:\n")
df.isnull().sum()



The number of missing values in each column of the dataset are as follows:



Country        0
City           0
Temperature    0
Feels_Like     0
Min_Temp       0
Max_Temp       0
Humidity       0
Pressure       0
Wind_Speed     0
Cloudiness     0
Description    0
Longitude      0
Latitude       0
Sunrise        0
Sunset         0
dtype: int64

### Multivariate Analysis:

#### Country-wise cities:

In [41]:
# check country and city column using groupby:

df.groupby('Country')['City'].nunique().sort_values(ascending=False)

Country
United States           9
China                   8
Italy                   5
Japan                   5
Spain                   5
Greece                  4
Canada                  3
Germany                 3
Mexico                  3
Turkey                  3
United Arab Emirates    3
France                  2
Australia               2
Saudi Arabia            2
Brazil                  2
United Kingdom          2
Vietnam                 2
Thailand                2
South Korea             2
Portugal                2
Israel                  2
Malaysia                2
India                   2
Poland                  2
Austria                 1
Argentina               1
Dominican Republic      1
Czech Republic          1
Denmark                 1
Belgium                 1
Chile                   1
Georgia                 1
Estonia                 1
Egypt                   1
Netherlands             1
Morocco                 1
Lithuania               1
Hungary                 1
Irel

#### City-wise temperature:

In [42]:
# check the temperature of each city:

df.groupby('City')['Temperature'].mean().sort_values(ascending=False)


City
Riyadh               40.08
Abu Dhabi            38.01
Mecca                37.22
Dubai                36.54
Marrakech            36.04
Muscat               36.01
Doha                 35.99
Sharjah              35.16
Cairo                34.42
Seville              33.18
Lisbon               32.43
Miami                31.25
Playa Del Carmen     31.01
Cancún               30.96
Tel Aviv             30.13
Antalya              29.49
Hong Kong            29.45
Zhuhai               29.28
Macau                29.20
Delhi                29.05
Hanoi                29.00
Shenzhen             28.97
Shanghai             28.92
Fukuoka              28.88
Punta Cana           28.87
Porto                28.82
Madrid               28.82
Islamabad            28.80
Osaka                28.47
Venice               28.45
Kyoto                28.30
Melbourne            28.10
Taipei               28.06
Guangzhou            27.97
Ho Chi Minh City     27.82
Bangkok              27.69
Orlando              27

#### City wise feels_like temperature:

In [50]:
#### check city with feels like temperature:

df.groupby('City')['Feels_Like'].mean().sort_values(ascending=False)

City
Muscat               42.72
Doha                 42.67
Dubai                41.84
Abu Dhabi            40.91
Sharjah              40.66
Miami                38.25
Playa Del Carmen     38.01
Cancún               37.96
Mecca                37.26
Riyadh               37.07
Zhuhai               36.28
Marrakech            36.22
Macau                36.20
Cairo                36.16
Hanoi                36.00
Punta Cana           35.87
Hong Kong            35.54
Shanghai             35.39
Fukuoka              35.27
Delhi                34.66
Shenzhen             34.66
Venice               34.20
Antalya              33.21
Tel Aviv             32.88
Kyoto                32.69
Seville              32.53
Melbourne            32.35
Osaka                32.25
Orlando              32.21
Ho Chi Minh City     31.37
Islamabad            31.15
Lisbon               30.68
Guangzhou            30.35
Bangkok              30.32
Singapore            29.93
Taipei               29.65
Mumbai               29

#### City-wise Min Tempearture:

In [52]:
# check city with Min temperature:

df.groupby('City')['Min_Temp'].mean().sort_values(ascending=False)

City
Riyadh               40.08
Abu Dhabi            38.01
Mecca                37.22
Marrakech            36.04
Muscat               36.01
Doha                 35.94
Sharjah              35.16
Dubai                35.14
Cairo                34.12
Seville              32.82
Playa Del Carmen     31.01
Cancún               30.59
Miami                30.56
Lisbon               30.51
Tel Aviv             29.60
Antalya              29.33
Delhi                29.05
Hanoi                29.00
Shanghai             28.92
Punta Cana           28.87
Macau                28.85
Zhuhai               28.84
Islamabad            28.80
Fukuoka              28.46
Madrid               28.34
Hong Kong            28.05
Kyoto                28.01
Shenzhen             27.99
Guangzhou            27.97
Osaka                27.51
Melbourne            27.21
Ho Chi Minh City     27.01
Mumbai               26.99
Bangkok              26.94
Taipei               26.85
Rhodes               26.81
Porto                26

#### City-wise Max Temperature:

In [51]:
# check city with Max temperature:

df.groupby('City')['Max_Temp'].mean().sort_values(ascending=False)

City
Riyadh               40.08
Marrakech            39.08
Abu Dhabi            38.06
Mecca                37.22
Sharjah              36.99
Dubai                36.96
Muscat               36.01
Doha                 35.99
Cairo                34.42
Lisbon               33.93
Seville              33.54
Miami                32.23
Porto                31.69
Playa Del Carmen     31.05
Cancún               30.96
Fukuoka              30.53
Macau                30.40
Zhuhai               30.40
Hong Kong            30.18
Tel Aviv             30.14
Shenzhen             30.12
Madrid               30.09
Venice               30.00
Barcelona            29.52
Antalya              29.49
Taipei               29.39
Delhi                29.05
Osaka                29.02
Hanoi                29.00
Melbourne            28.99
Shanghai             28.93
Punta Cana           28.87
Islamabad            28.80
Kyoto                28.71
Orlando              28.35
Tokyo                28.12
Athens               28

#### City-wise Humidity:

In [43]:
# check the humidity of each city:

df.groupby('City')['Humidity'].mean().sort_values(ascending=False)

City
Vancouver            97.0
Phuket               94.0
Prague               92.0
Budapest             91.0
San Francisco        91.0
Punta Cana           89.0
Tallinn              89.0
Tokyo                89.0
Vienna               89.0
Hanoi                89.0
Lima                 88.0
Orlando              88.0
Johor Bahru          87.0
Guilin               87.0
Dublin               85.0
Sapporo              85.0
Venice               85.0
Fukuoka              84.0
Macau                84.0
Shanghai             84.0
Zhuhai               84.0
Toronto              83.0
Busan                83.0
Singapore            83.0
Los Angeles          82.0
Mexico City          82.0
Melbourne            80.0
Shenzhen             80.0
Kuala Lumpur         80.0
Kyoto                79.0
Playa Del Carmen     79.0
Delhi                79.0
Cancún               79.0
Honolulu             79.0
Miami                79.0
Hong Kong            78.0
Ho Chi Minh City     78.0
Mumbai               78.0
São Pau

#### City-wise Cloudiness:

In [48]:
# check the cloudiness of each city:

df.groupby('City')['Cloudiness'].mean().sort_values(ascending=False)

City
Hanoi                100.0
Bangkok              100.0
Lima                 100.0
Kyoto                100.0
Rome                 100.0
Tallinn              100.0
Venice               100.0
Prague               100.0
London               100.0
Los Angeles          100.0
Kuala Lumpur          98.0
Munich                98.0
Osaka                 95.0
Muscat                90.0
Johor Bahru           75.0
Busan                 75.0
Delhi                 75.0
Istanbul              75.0
San Francisco         75.0
Shanghai              75.0
Punta Cana            75.0
Mumbai                75.0
Miami                 75.0
Mexico City           75.0
Singapore             75.0
São Paulo             75.0
Budapest              75.0
Cairo                 75.0
Tokyo                 75.0
Vancouver             75.0
Vienna                75.0
Rio de Janeiro        75.0
Guilin                68.0
Fukuoka               48.0
Antalya               40.0
Amsterdam             40.0
Warsaw                4

In [54]:
df.groupby('City')['Pressure'].mean().sort_values(ascending=False)

City
New York City        1028.0
Sydney               1028.0
Montreal             1028.0
Brussels             1026.0
Paris                1026.0
Toronto              1026.0
London               1026.0
Amsterdam            1025.0
Buenos Aires         1025.0
Munich               1024.0
Stockholm            1024.0
Copenhagen           1024.0
Frankfurt            1023.0
Tallinn              1023.0
Zurich               1022.0
São Paulo            1021.0
Santiago             1020.0
Edinburgh            1020.0
Berlin               1020.0
Rio de Janeiro       1020.0
Sapporo              1019.0
Rome                 1018.0
Prague               1018.0
Vilnius              1017.0
Mexico City          1017.0
Madrid               1017.0
Palma de Mallorca    1017.0
Porto                1017.0
Barcelona            1016.0
Vienna               1015.0
Valencia             1015.0
Vancouver            1015.0
Beijing              1015.0
Lisbon               1015.0
Kyoto                1015.0
Punta Cana     

### **Part 3: Data Cleaning**


#### Step 1: Delete Unnecessary Columns and Null/Error Rows




In [57]:
# remove the duplicated data in the dataset:

df.drop_duplicates(inplace=True)

In [58]:
df.shape


(100, 15)

>It shows that there were no duplicates in the dataset.

#### Step 2: Handle Missing or Erroneous Data

In [59]:

# Check for missing values
print(df.isnull().sum())

# Drop rows with any null values in the dataset
df.dropna(inplace=True)


Country        0
City           0
Temperature    0
Feels_Like     0
Min_Temp       0
Max_Temp       0
Humidity       0
Pressure       0
Wind_Speed     0
Cloudiness     0
Description    0
Longitude      0
Latitude       0
Sunrise        0
Sunset         0
dtype: int64


In [60]:
df.head()

# the dataset seems better now.

Country       City  Temperature  Feels_Like  Min_Temp  \
0                France      Paris        19.58       18.79     18.45   
1  United Arab Emirates      Dubai        36.54       41.84     35.14   
2                 Spain     Madrid        28.82       27.24     28.34   
3                 Japan      Tokyo        26.64       26.64     25.34   
4           Netherlands  Amsterdam        18.60       18.27     17.69   

   Max_Temp  Humidity  Pressure  Wind_Speed  Cloudiness  \
0     20.70        46      1026        2.57           0   
1     36.96        44      1002        8.23           0   
2     30.09        15      1017        2.06           0   
3     28.12        89      1013        4.12          75   
4     19.38        67      1025        4.63          40   

                   Description  Longitude  Latitude   Sunrise    Sunset  
0                    Clear sky     2.3488   48.8534  10:27 AM  11:04 PM  
1                    Clear sky    55.3047   25.2582  07:04 AM  07:23 PM  
2                    Clear sky    -3.7026   40.4165  10:55 AM  11:24 PM  
3  Light intensity shower rain   139.6917   35.6895  01:23 AM  01:49 PM  
4             Scattered clouds     4.8897   52.3740  10:15 AM  10:56 PM

#### Step 3: Normalize Column Names

In [ ]:

# Normalize column names (lowercase and replace spaces with underscores)
df.columns = df.columns.str.lower()

#### Step 4: Remove Duplicate Rows


In [61]:

# Remove duplicate rows
df.drop_duplicates(inplace=True)

# Insights: 
> 1. There are `100 rows` and `15 columns` in the dataset.
2. Number of unique values in the column 'Country' are: `47`
3. Number of unique values in the column 'City' are: `100`
4. Number of unique values in the column 'Description' are: `12`
5. Number of unique values in the column 'Sunrise' are: `91`
6. Number of unique values in the column 'Sunset' are: `94`
> 7. there are `no duplicates` in the dataset.
1. The minimum value of Temperature is `7.24` 
2. The maximum value of Temperature is `34.38`
3.  The minimum value of Feels_Like is `4.38` 
4.  The maximum value of Feels_Like is `41.0`
5.  The minimum value of Min_Temp is `6.44`
    - The city is `Sydney` 
6.  The maximum value of Min_Temp is `34.01`
7.  The minimum value of Max_Temp is `8.51` 
8.  The maximum value of Max_Temp is `40.08`
    - The city is `Riyadh`
9.  The minimum value of Humidity is `10`
    - The city is `Verona` 
10. The maximum value of Humidity is `97.0`
    - The city is `Vancouver`
11. The minimum value of Pressure is `1000` 
12. The maximum value of Pressure is `1030`
13. The minimum value of Wind_Speed is `0.51` 
14. The maximum value of Wind_Speed is `12.35`
15. The minimum value of Cloudiness is `0` 
    1.  There are a `10 cities` which has the `0` cloudiness
16. The maximum value of Cloudiness is `100`
    1.  There are `30 cities` which has the `100` cloudiness
17. The minimum value of Longitude is `-157.8583` 
18. The maximum value of Longitude is `151.2073`
19. The minimum value of Latitude is `-34.6132` 
20. The maximum value of Latitude is `59.437`
> 27. There are no missing values in the dataset
